Build official lustre tags

dinatale2 opened the pull request:
Created a git tag polling class to collect new tags from a git repository and submit them as changes to the build master. Enabled the lustre buildbot to use the new polling class to collect and build official tags from the lustre git repository.

morrone commented on master/lustregittagpoller.py:
```diff
@@ -0,0 +1,224 @@
+                 pollinterval=-2, pollAtLaunch=False,
+                 encoding='utf-8', usetimestamps=True):
+
+        if pollinterval != -2:
```
I have absolutely no idea what you are trying to accomplish here.  Neither will anyone else, because there are no comments explaining this.  But even without knowing what is going on, I can tell you that it is not at all wise to have two options that are named nearly identically, differing only in the capitalization of a single letter in the middle.

dinatale2 replied:
The gitpoller class has pollinterval for legacy reasons. This class is a new class so perhaps I shouldn't support pollinterval?

morrone commented on master/lustregittagpoller.py:
```diff
@@ -0,0 +1,224 @@
+
+class LustreTagPoller(base.PollingChangeSource, StateMixin):
+
+    """ This source will poll a remote git repo for
```
Please read the python [Docstring Conventions](https://www.python.org/dev/peps/pep-0257/) and abide by them.  More explanation is definitely called for in this instance.

morrone replied:
The first line of the comment should not span two lines:

"Multi-line docstrings consist of a summary line just like a one-line docstring, followed by a blank line, followed by a more elaborate description. The summary line may be used by automatic indexing tools; it is important that it fits on one line and is separated from the rest of the docstring by a blank line."

dinatale2 commented:
Updated. Note there are two commits in this PR, the comments were changed in those.

morrone commented:
OK, but you still didn't really explain the design approach in the commit comment.  It seems like there is an awful lot of cut-and-paste from GitPoller.  Explain why doing that is the right approach rather than sub-classing GitPoller.  It isn't clear to me why we should have all this added complexity of using defer, only to turn around and inline all of the defers.  Doesn't that defeat the purpose of having defers?  Maybe there was a reason for it in the original class, but I'm not seeing it here?

And if the cut-and-paste makes sense rather than subclassing, I'd like to see the code cleaned up and customized for our specific purpose.

Also, please update the commit messages to be present tense.

morrone commented on master/lustregittagpoller.py:
```diff
@@ -0,0 +1,246 @@
+            repository=self.repourl,
+            src='git')
+
+    # originally from gitpoller
```
Python is a case-sensitive language.  The class name is GitPoller, not gitpoller.

morrone commented on master/master.cfg:
```diff
@@ -287,6 +294,13 @@ c['schedulers'] = [
+        name="tag-changes",
+        change_filter=util.ChangeFilter(
+            project=gerrit_project,
+            branch_re="^refs/tags/[0-9]+\.[0-9]+\.[0-9]+$"),
```
This regex is too restrictive.  See Documentation/versioning.txt on the Lustre master branch.

dinatale2 commented:
Updated.

morrone commented on master/lustregittagpoller.py:
```diff
@@ -0,0 +1,122 @@
+            yield self._dovccmd('init', ['--bare', self.workdir])
+
+        log.msg("LustreGitTagPoller: collecting references")
+        newRefs = yield self._getRefs()
```
ls-remote is an over-the-network operation.  After the very next step, the "fetch --tags", the information we need will be available locally, correct?  Can't we skip the ls-remote?

dinatale2 commented:
Updated. Changed ls-remote to show-ref and fixed how the output is parsed. Also changed the poll function to perform a git fetch --tags before the show-ref. That should make all those operations local.

morrone commented on master/lustregittagpoller.py:
```diff
@@ -0,0 +1,118 @@
+        return str
+
+    def _getRefs(self):
+        """Collect and parse all remote references"""
```
Function comment needs updating.

morrone commented on master/lustregittagpoller.py:
```diff
@@ -0,0 +1,118 @@
+
+    @defer.inlineCallbacks
+    def _process_change(self, ref, rev):
+        """Process each tag """
```
This comment is not terribly useful.  What does it mean to "process" a tag?  What exactly is the process that is being applied?  A tag goes in, what comes out?

morrone commented on master/lustregittagpoller.py:
```diff
@@ -0,0 +1,118 @@
+
+    @defer.inlineCallbacks
+    def poll(self):
+        log.msg("LustreGitTagPoller: poll iteration started")
```
These messages don't match the name of the class.

morrone commented on master/lustregittagpoller.py:
```diff
@@ -0,0 +1,118 @@
+
+    This source will poll a remote git repository for new tags and finds changes
+    to existing tags. Once a new or changed tag is found, a change is submitted to
+    the build master.""" 
```
This needs more detail.  How do you determine which tags are "new" or "changed"?  How does this persistently track the list of seen tags?

dinatale2 commented:
Updated all the comments you suggested.

morrone commented on master/lustregittagpoller.py:
```diff
@@ -0,0 +1,122 @@
+    checked against the prior set. A tag is considered new if it doesn't exist in
+    the set. A tag is considered change when the tag has been seen but its revision
+    has changed. The set is maintain in the buildbot master's database."""
+
```
Need information about arguments too.  Note that this only accepts keyword arguments, and that the keyword arguments are the same as the arguments to the parent GitPoller class.  Note any exceptions (see next comment in _init_)

morrone commented on master/lustregittagpoller.py:
```diff
@@ -0,0 +1,122 @@
+    def __init__(self, **kwargs):
+        if kwargs.get('workdir') is None:
+            kwargs['workdir'] = 'lustretagpoller-work'
+
```
Some of the arguments in the parent class no longer make sense, and should be caught here, and an exception raised.  "branch" and "branches" are invalid options to this subclass, right?  Go through the rest of the options and see if any others are now invalid.  Note exceptions in the comment above.

morrone commented on master/lustregittagpoller.py:
```diff
@@ -0,0 +1,122 @@
+from buildbot.util import epoch2datetime
+from buildbot.changes.gitpoller import GitPoller 
+
+class LustreTagPoller(GitPoller):
```
Since you are making the "branch" and "branches" arguments invalid in this class, you will also need to override the parent class's describe() method with your own in this subclass.  Look through the parent class for other methods that need to be replaced to make sense in this subclass.

dinatale2 commented:
Updated. The branches and branch arguments appear to be the only ones we should issue an error on when initializing a lustretagpoller. The poll and describe functions were previously overridden and I don't think other functions need an override. I implemented a process_change function instead of overriding process_changes from the gitpoller class for naming reasons. Also updated the docstring for the class.

morrone commented:
OK, I think that looks good now.  Just a minor point for the future:  using "in" and "not in" in if-statements is a bit more pythonic than using .get() and comparing against None.
